### PR TITLE
PGF: Consistently set LaTeX document font size

### LINKED ISSUE #26892: [Bug]: PGF font size mismatch between measurement and output
### Bug summary

When using the PGF backend, different LaTeX document font size settings are used when measuring widths vs. producing output. Although this is usually not a problem (since all text produced by the PGF backend locally sets the font properties), packages used in the custom preamble (such as `unicode-math`) may depend on the font size settings, leading to incorrect width measurements since different outcomes are produced in the two cases.

### Code for reproduction

```python
import matplotlib
import matplotlib.pyplot as plt

matplotlib.use('pgf')

plt.rc('text', usetex=True)  # not sure if this is actually relevant for PGF backend?
plt.rc('pgf', texsystem='lualatex', rcfonts=False, preamble=r'''
    \usepackage{unicode-math}
''')

plt.figure()
plt.plot([], label=r'''$this is a very very very long math label a \times b + 10^{-3}$ and some text''')
plt.legend()
plt.savefig('test_math.png')
plt.savefig('test_math.pgf')
```

### Actual outcome

Math text has a different width (characters are comparatively wider when scaled to the same height) and thus exceeds the legend boundary (tick labels also have some slight spacing issues):

MWE output showing the problem

### Expected outcome

Text should be within the legend boundary:

Fixed MWE output

### Additional information

The main issue causing the problem in my example is that different font sizes are passed to `\documentclass` in different cases:

https://github.com/matplotlib/matplotlib/blob/ff552f45f2d30c76120fe0355306afac9a9dd2da/lib/matplotlib/backends/backend_pgf.py#L186-L188

https://github.com/matplotlib/matplotlib/blob/ff552f45f2d30c76120fe0355306afac9a9dd2da/lib/matplotlib/backends/backend_pgf.py#L813

https://github.com/matplotlib/matplotlib/blob/ff552f45f2d30c76120fe0355306afac9a9dd2da/lib/matplotlib/backends/backend_pgf.py#L923

This specific issue is only present in matplotlib ≥ v3.6.0 seems to be a historical accident carried over from when the document class was changed from `minimal` to `article` in 4d85f703611 (#23443). `\documentclass{minimal}` actually ignores any options passed to it (including `12pt`), so that option in the original code was spurious and there was no mismatch between the cases with and without this option. But when this was changed to `\documentclass{article}`, the option `12pt` did start taking effect, leading to the discrepancy.

However, more generally, the code always had an implicit dependence on the LaTeX font size setting, which does affect things like the `unicode-math` package (which selects different versions of math fonts depending on the font size). So when saving as `.pgf` and using these files directly in LaTeX documents, there was still the potential for problems to show up. Another consequence of this is that the standard LaTeX font commands (`\tiny`, …, `\normalsize`, …, `\Huge`) were also relative to this obscure font size setting (instead of `rcParams['font.size']`).

**Workaround for current releases:** Manually set the font size in the preamble.

```python
import matplotlib
import matplotlib.pyplot as plt

matplotlib.use('pgf')

plt.rc('text', usetex=True)  # not sure if this is actually relevant for PGF backend?
plt.rc('pgf', texsystem='lualatex', rcfonts=False, preamble=r'''
    \usepackage[fontsize=10pt]{scrextend}
    \usepackage{unicode-math}
''')

plt.figure()
plt.plot([], label=r'''$this is a very very very long math label a \times b + 10^{-3}$ and some text''')
plt.legend()
plt.savefig('test_math.png')
plt.savefig('test_math.pgf')
```

### Operating system

_No response_

### Matplotlib Version

3.8.0

### Matplotlib Backend

pgf

### Python version

_No response_

### Jupyter version

_No response_

### Installation

pip

### PR DESCRIPTION
**Pull request by Socob**
## PR summary

Currently, when using the PGF backend, different LaTeX document font size settings are used when measuring widths vs. producing output. Although this is usually not a problem (since all text produced by the PGF backend locally sets the font properties), packages used in the custom preamble (such as `unicode-math`) may depend on the font size settings, leading to incorrect width measurements since different outcomes are produced in the two cases. This fixes the issue by always setting the LaTeX font size to the rc setting `font.size`.

Closes #26892.

## PR checklist

- [x] "closes #0000" is in the body of the PR description to [link the related issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)
- [x] new and changed code is [tested](https://matplotlib.org/devdocs/devel/testing.html)
- [N/A] *Plotting related* features are demonstrated in an [example](https://matplotlib.org/devdocs/devel/document.html#write-examples-and-tutorials)
- [N/A] *New Features* and *API Changes* are noted with a [directive and release note](https://matplotlib.org/devdocs/devel/coding_guide.html#new-features-and-api-changes)
- [N/A] Documentation complies with [general](https://matplotlib.org/devdocs/devel/document.html#write-rest-pages) and [docstring](https://matplotlib.org/devdocs/devel/document.html#write-docstrings) guidelines

### CODE REVIEW

**Comment by Socob** [comment]
Not sure what’s going on with the flake8 test thing – it doesn’t look like I changed anything about blank lines where it’s complaining.

File: lib/matplotlib/backends/backend_pgf.py [hunk 1]
```diff
@@ -23,37 +23,55 @@
...
 from matplotlib.path import Path
 from matplotlib.figure import Figure
+from matplotlib.font_manager import FontProperties
 from matplotlib._pylab_helpers import Gcf
 
 _log = logging.getLogger(__name__)
 
 
+DOCUMENTCLASS = r"\documentclass{article}"
+
```
Comment (oscargus) [inline lib/matplotlib/backends/backend_pgf.py:33]: Agreed that the linting rule is non-obvious here.

However, I guess that adding an additional empty line here should most likely solve it.

Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py:33]: Hm OK, I didn’t make the connection to that line since it mentioned line 38, but I can try if adding a line there changes anything.

**Review by oscargus — APPROVED** [review]
I think this looks good!

However, it may be worthwhile adding an image test to confirm that the new behavior is kept.

Also, it may be that one will need to install `scrextend` in the CI to get it going (not sure if it is included in any standard distribution or not).

Finally, it may be helpful to add a note that this will work better if `scrextend` is available. Maybe as an additional bullet at the end of https://github.com/matplotlib/matplotlib/blob/main/galleries/users_explain/text/pgf.py
(https://matplotlib.org/stable/users/explain/text/pgf.html)

**Comment by Socob** [comment]
> However, it may be worthwhile adding an image test to confirm that the new behavior is kept.

Sounds like a good idea – I’ll get on that.

> Also, it may be that one will need to install `scrextend` in the CI to get it going (not sure if it is included in any standard distribution or not).

`scrextend` is part of the KOMA script bundle, which is very widely used. In TeX Live, it’s part of `collection-latexrecommended`, the same as `fontspec` and `underscore` which are currently already loaded unconditionally. So normally, it should be present if the other two are.

Either way, the use of `scrextend` is really just to set up the font size commands `\tiny`, ..., `\normalsize`, ..., `\Huge` with sensible values. I don’t think these are commonly used, and the fallback takes care of the one that really matters when it comes to this issue (`\normalsize`).

> Finally, it may be helpful to add a note that this will work better if `scrextend` is available. Maybe as an additional bullet at the end of https://github.com/matplotlib/matplotlib/blob/main/galleries/users_explain/text/pgf.py (https://matplotlib.org/stable/users/explain/text/pgf.html)

Not sure if this is necessary given the use of `fontspec` and `underscore` (see above)? The only reason I added the `\IfFileExists` check is really just because `scrextend` is not absolutely *necessary* to fix the main issue, and I wanted to assume as little as possible about the user’s TeX installation.

**Comment by oscargus** [comment]
Then it should be fine I think! 

(Or rather, it would be nice that the required packages are explicitly stated somewhere, I may have missed it though, but this PR doesn't strictly add something to that.)

**Comment by tacaswell** [comment]
@pwuertz Do you have bandwidth to review this?

File: lib/matplotlib/backends/backend_pgf.py [hunk 2]
```diff
@@ -23,37 +23,52 @@
     _create_pdf_info_dict, _datetime_to_pdf)
 from matplotlib.path import Path
 from matplotlib.figure import Figure
+from matplotlib.font_manager import FontProperties
 from matplotlib._pylab_helpers import Gcf
 
 _log = logging.getLogger(__name__)
 
 
+DOCUMENTCLASS = r"\documentclass{article}"
```
Comment (timhoffm) [inline lib/matplotlib/backends/backend_pgf.py:32]: Why do you want to pull this out:
- should this really be public?
- do you anticipate to modify this?

Comment (voidstarstar) [reply lib/matplotlib/backends/backend_pgf.py:32]: @timhoffm There may be a good reason for a user to want to modify this: https://github.com/matplotlib/matplotlib/issues/28119

Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py:32]: @timhoffm Sorry, I missed this comment. The main reason why I pulled it out into a constant is because inconsistent documentclass declarations are precisely what caused the issue that this is fixing. I’m not sure what you mean by “public” – do you want the variable name to start with an underscore?

Comment (timhoffm) [reply lib/matplotlib/backends/backend_pgf.py:32]: Yes, please use an underscore if it's not meant to be modified by users. We're quite strict on this nowadays because otherwise backward-compatible changes become very hard.

If you want users to give the freedom to change `DOCUMENTCLASS`, then it would need documentation.

Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py]: OK, done!

**Comment by Socob** [comment]
I’ve made the last requested change and rebased. So presumably it’s ready to merge?

File: lib/matplotlib/backends/backend_pgf.py [hunk 3]
```diff
@@ -23,37 +23,52 @@
...
+        size=mpl.rcParams["font.size"]
+    ).get_size_in_points()
     return "\n".join([
         # Remove Matplotlib's custom command \mathdefault.  (Not using
         # \mathnormal instead since this looks odd with Computer Modern.)
         r"\def\mathdefault#1{#1}",
         # Use displaystyle for all math.
         r"\everymath=\expandafter{\the\everymath\displaystyle}",
+        # Set up font sizes to match font.size setting.
+        r"\IfFileExists{scrextend.sty}{",
```
Comment (anntzer) [inline lib/matplotlib/backends/backend_pgf.py:51]: Please document that (IIUC?) the effecf of scrextend (koma-script) here is that all fontsize selectors (HUGE, Large, large, etc.) will be correctly sized as well.

File: lib/matplotlib/backends/backend_pgf.py [hunk 4]
```diff
@@ -94,6 +109,8 @@ def _escape_and_apply_props(s, prop):
     family = prop.get_family()[0]
     if family in families:
         commands.append(families[family])
+    elif not mpl.rcParams["pgf.rcfonts"]:
+        commands.append(r"\rmfamily")
```
Comment (voidstarstar) [inline lib/matplotlib/backends/backend_pgf.py:113]: Is defaulting to serif documented anywhere?  I see one page that says that if `pgf.rcfonts` is disabled, then you will "do the font configuration yourself", which I assumed meant in the preamble.
https://github.com/matplotlib/matplotlib/blob/eb62d6951a5d7657fe006bb919cbae08481023e2/galleries/users_explain/text/pgf.py#L122
For example, will this break expected behavior if a user preamble says `\renewcommand{\familydefault}{\sfdefault}`?

**Comment by voidstarstar** [comment]
I think that the `documentclass` is also set in `texmanager.py` for use with the `usetex` flag.  In the future, should `_DOCUMENTCLASS` cover this as well?
https://github.com/matplotlib/matplotlib/blob/eb62d6951a5d7657fe006bb919cbae08481023e2/lib/matplotlib/texmanager.py#L203

File: lib/matplotlib/backends/backend_pgf.py (hunk 3 above)
Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py:51]: Yes, when passing the package option `fontsize=`, that’s the effect of `scrextend`. Is a comment enough or do you mean user-facing documentation?

Comment (anntzer) [reply lib/matplotlib/backends/backend_pgf.py:51]: A comment is enough.

File: lib/matplotlib/backends/backend_pgf.py (hunk 4 above)
Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py:113]: I don’t think it’s documented anywhere. I think I made the change to be consistent with the other branches (which always set a font family), and to protect against outer-level font changes messing with the figure fonts when the figure is included in a TeX document. `\fontfamily{\familydefault}\selectfont` may be a better choice?

**Comment by Socob** [comment]
Since `texmanager.py` is used for a completely different purpose, I didn’t touch it. I also don’t think that `texmanager.py` and PGF should necessarily use the same `\documentclass`.

File: lib/matplotlib/backends/backend_pgf.py (hunk 4 above)
Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py]: For the moment I changed it to `\fontfamily{\familydefault}`. If anyone else has any opinions about this, they can always speak up ;-)

File: lib/matplotlib/backends/backend_pgf.py (hunk 3 above)
Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py:54]: Done!

**Comment by Socob** [comment]
Once again added all requested changes!

**Comment by Socob** [comment]
Is there anything else that needs to be done? I don’t know what the codecov test is complaining about, nothing in there looks related to my changes.

File: lib/matplotlib/backends/backend_pgf.py [hunk 5]
```diff
@@ -23,37 +23,55 @@
...
+        r"  \renewcommand{\normalsize}{\fontsize{%f}{%f}\selectfont}"
+        % (font_size_pt, 1.2 * font_size_pt),
+        r"  \normalsize",
+        r"}",
         # Allow pgf.preamble to override the above definitions.
         mpl.rcParams["pgf.preamble"],
-        r"\ifdefined\pdftexversion\else  % non-pdftex case.",
-        r"  \usepackage{fontspec}",
         *([
+            r"\ifdefined\pdftexversion\else  % non-pdftex case.",
```
Comment (anntzer) [inline lib/matplotlib/backends/backend_pgf.py:64]: Is there a reason for the change here and immediately below?

File: lib/matplotlib/backends/backend_pgf.py [hunk 6]
```diff
@@ -23,37 +23,55 @@
...
             r"  \%s{%s}[Path=\detokenize{%s/}]"
             % (command, path.name, path.parent.as_posix())
             for command, path in zip(
                 ["setmainfont", "setsansfont", "setmonofont"],
                 [pathlib.Path(fm.findfont(family))
                  for family in ["serif", "sans\\-serif", "monospace"]]
             )
-        ] if mpl.rcParams["pgf.rcfonts"] else []),
-        r"\fi",
+        ] + [r"\fi"] if mpl.rcParams["pgf.rcfonts"] else []),
```
Comment (anntzer) [inline lib/matplotlib/backends/backend_pgf.py:74]: ... i.e. here?

File: lib/matplotlib/backends/backend_pgf.py [hunk 7]
```diff
@@ -185,7 +205,7 @@ class LatexManager:
     @staticmethod
     def _build_latex_header():
         latex_header = [
-            r"\documentclass{article}",
```
Comment (anntzer) [inline lib/matplotlib/backends/backend_pgf.py:188]: I think these could have been left as they were originally (but removing the 12pt below) no?  especially considering that we may want to make the documentclass configurable in the future.

File: lib/matplotlib/tests/test_backend_pgf.py [hunk 8]
```diff
@@ -406,3 +406,27 @@ def test_sketch_params():
...
+
+
+# test to make sure that the document font size is set consistently (see #26892)
+@needs_pgf_xelatex
+@pytest.mark.skipif(
+    not _has_tex_package('unicode-math'), reason='needs unicode-math.sty'
+)
+@pytest.mark.backend('pgf')
+@image_comparison(['pgf_document_font_size.pdf'], style='default', remove_text=True)
+def test_document_font_size():
```
Comment (anntzer) [inline lib/matplotlib/tests/test_backend_pgf.py:418]: 1) do you really need unicode-math for this test?  I'd say it's not particularly relevant, no?
2) you could do without the plot at all and simply use figtext() to draw text on an empty figure.

Comment (Socob) [reply lib/matplotlib/tests/test_backend_pgf.py:418]: 1. Not `unicode-math` specifically, no. But if I remember correctly, `unicode-math` (and other math font packages) select the font variant they use at package load time based on the document font size, and these variants have different metrics even for the same font size, which is what caused the original issue. So I’d either need `unicode-math` or some random font package with the same behavior, and `unicode-math` seems much more universal (especially since matplotlib already uses it by default).
2. I suppose I could. I just took the easiest path and used the example I already had in #26892. Let me know if you really want me to make and test a new example.

File: lib/matplotlib/backends/backend_pgf.py (hunk 5 above)
Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py:64]: That’s 14a59001. I guess it’s only tangentially related to the rest of this PR, but I noticed that `\usepackage{fontspec}` was always included in the preamble, which is not necessarily desirable when doing custom font setup (i. e. `pgf.rcfonts=False`).

File: lib/matplotlib/backends/backend_pgf.py (hunk 6 above)
Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py:74]: See above.

File: lib/matplotlib/backends/backend_pgf.py (hunk 7 above)
Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py:188]: Not sure what you mean? These all have to be identical, so introducing a single name for them is just DRY. And in fact one of the first things one would do when making the documentclass configurable. I don’t see how it matters whether it’s a global constant or something else – that can always be changed with no effort. The point is to use a single variable for things that have to be the same.

**Comment by voidstarstar** [comment]
I think this PR tries to fix too much.  Isn't [fixing the 12pt mismatch](https://github.com/matplotlib/matplotlib/pull/26893/commits/46ac87e417ec4141ece678884bfd739328ce5712) the only thing necessary to fix the original bug demonstrated by the code for reproduction?  Everything else seems to target a slightly different bug (or bugs) that I do not know how to reproduce, but they sound to me like they might be solved by setting a custom `documentclass` (see https://github.com/matplotlib/matplotlib/issues/28119).


Maybe I'm wrong, but I thought that the `font.size` rcParam option was supposed to serve a different purpose than the global document text size and I'm now wondering if this PR confuses the two.  I thought that `font.size` was the default size for the figure and that matplotlib provides [special sizes](https://github.com/matplotlib/matplotlib/blob/378a77a339a9ba30551900482d8bb598344df6fc/lib/matplotlib/mpl-data/matplotlibrc#L253-L260) relative to `font.size` for this purpose already.  Instead of basing the LaTeX font size commands (`\tiny`, etc.) on `font.size`, wouldn't a better solution be to just allow the user to set their own custom `documentclass` (which can include their own global document text size)?  I think this would fix the implicit dependence on the size problem mentioned in the original issue.

In other words, if the user could set the `documentclass` in matplotlib to match their own custom `documentclass`, wouldn't that fix this issue in a better way without needing to force the global document font size to match `font.size`?

**Comment by Socob** [comment]
I really didn’t expect “tries to fix too much” on a PR that adds/changes 21 lines of code (excluding comments/tests/blanks)! I don’t know what you mean by “everything else” beyond adjusting the LaTeX font commands?

Since there is currently no way to set the documentclass, `font.size` is the only available information on what the user’s document font size is. The matplotlib internal sizes are probably preferable in some cases (when they can be used), but since I was setting `\normalsize`, I thought it’d be more consistent to adjust the other LaTeX font size commands accordingly (this is 3 lines of code in the PR). Since there is no documentation on this, users will expect that these font commands will work “as expected”, as evidenced by [posts such as this one](https://stackoverflow.com/a/26769919).

Sure, making the documentclass configurable may be a better way to fix this than my (minimal!) change. Feel free to write that PR!

**Comment by Socob** [comment]
@anntzer Do my answers resolve your review comments or do you want me to change anything about the PR?

File: lib/matplotlib/backends/backend_pgf.py (hunk 3 above)
Comment (voidstarstar) [reply lib/matplotlib/backends/backend_pgf.py:54]: Is there a reason why this `scrextend` code is not guarded by the `pgf.rcfonts` flag too?  I think the `scrextend` change is the only thing that I had some opposition to because, unless I'm mistaken, I think it can break consistency.  If it's only used when `pgf.rcfonts=False`, then I think it would at least have an easy workaround.

Comment (Socob) [reply lib/matplotlib/backends/backend_pgf.py:54]: It seems more consistent to me to always set the document font size to a well-defined value, rather than having this depend on `pgf.rcfonts`. What do you mean by “break consistency”? There shouldn’t be any difference in plot outputs, since font sizes are always set locally as well.

File: lib/matplotlib/tests/test_backend_pgf.py (hunk 8 above)
Comment (Socob) [reply lib/matplotlib/tests/test_backend_pgf.py:418]: Actually, I think I was mistaken with my comment that matplotlib uses `unicode-math` (I was thinking of `fontspec`). Nevertheless, the rest still applies.

**Comment by voidstarstar** [comment]
> I really didn’t expect “tries to fix too much” on a PR that adds/changes 21 lines of code (excluding comments/tests/blanks)! I don’t know what you mean by “everything else” beyond adjusting the LaTeX font commands?

Yes, the "everything else" are the other font related changes (fixing a scope bug by defaulting to `\familydefault`, fixing `fontspec` to depend on `pgf.rcfonts`, and using `scrextend` to set relative text sizes).

To be clear, I wasn't saying that the changes are bad or that they should not be merged.  I actually think most are strong improvements and a dependency for a PR I'm currently writing too.  I just did not fully understand the motivation behind the `scrextend` related code because I think in some cases it might *introduce* a consistency issue since it is a global setting.  Thank you for the SO link, it helps me better understand the use case.

**Comment by voidstarstar** [comment]
The following is a proof of concept of the possible consistency issue I mentioned:

LaTeX source:
```LaTeX
\documentclass[12pt]{article}
\usepackage{pgf}
\begin{document}
\Huge Before
\input{plot.pgf}
\Huge After
\end{document}
```

Python source:
```Python
import matplotlib
matplotlib.use("pgf")
import matplotlib.pyplot as plt

# Set up PGF
pgf_with_latex = {
    "pgf.texsystem": "lualatex",
    "font.family": "serif",
    "text.usetex": True,
    "pgf.rcfonts": False,
    "font.size": 1,
    "figure.autolayout": True,
    "pgf.documentclass": r"\documentclass[12pt]{article}",                                                                                                                                                         
    
    "axes.labelsize": "xx-small",
}
matplotlib.rcParams.update(pgf_with_latex)

# Create some sample data
x = [0, 1, 2, 3, 4]
y = [0, 0.2, 0.4, 0.1, 0.3]

# Create the plot
fig, ax = plt.subplots(figsize=(4.5, 3))
ax.plot(x, y)

ax.set_title(r"Sample Plot")

ax.set_xlabel(r"\Huge X")

ax.set_ylabel("Y")

# Save the plot
plt.savefig("plot.pgf")
plt.savefig("plot.png", dpi=300)
```

I think the above example might demonstrate the problem I was alluding to.  I use extremely big/small sizes to make the problem more obvious visually.  The `font.size` is set to 1.  This causes `scrextend` to change `\Huge` to be much smaller (since it's relative to 1 instead of 12).  Matplotlib then internally renders the xlabel `"r\Huge X"` which is rendered small as expected.  The problem arises when the pgf file is imported into LaTeX.  The LaTeX document preamble does not include `scrextend`, so the xlabel is much bigger than expected and therefore is misaligned.  We can't include `scrextend` in the preamble because it changes the font globally, which will affect the other text in the document.

My recommended solution would be to leave the `\normalsize` and other commands (`\tiny`, etc.) alone to match the LaTeX file, or if you must change them using `scrextend`, do so using `pgf.preamble`.  Ideally, I think the plot should not rely on changing any global settings.

> Sure, making the documentclass configurable may be a better way to fix this than my (minimal!) change. Feel free to write that PR!

I'm currently working on #28167 for this.

File: lib/matplotlib/backends/backend_pgf.py (hunk 3 above)
Comment (voidstarstar) [reply lib/matplotlib/backends/backend_pgf.py:54]: I provided an example here due to space limitations: https://github.com/matplotlib/matplotlib/pull/26893#issuecomment-2094003950

Please let me know if I made a mistake.  It's very likely that I am just misunderstanding something and this is a non-issue.

**Comment by anntzer** [comment]
Thank you @voidstarstar for your example.  I agree we can't assume that the user will indeed be using koma-script just because it happens to be present in the install (if I understood the problem you're mentioning correctly).

**Comment by Socob** [comment]
I think there is a lot of confusion going on here. First of all, `\usepackage[fontsize=<N>pt]{scrextend}` has nothing to do with whether or not the user is using KOMA-Script. It simply replicates what would happen when setting the document font size, as in `\documentclass[<N>pt]{article}`. Now, why did I not simply pass the font size to the documentclass? Because the standard LaTeX classes, like `article`, don’t support arbitrary font sizes, wheres KOMA-Script does. Since the user cannot set the documentclass, this is the next-best thing to support as many cases as possible. (And for obvious reasons, I didn’t want to change the documentclass used by the PGF backend in this PR.)

Now, for @voidstarstar’s example: Of course, the assumption is that **the user will set `font.size` to match the font size of their document**! What scenario or use case would there be where any other choice makes sense?

> My recommended solution would be to leave the `\normalsize` and other commands (`\tiny`, etc.) alone to match the LaTeX file

How? matplotlib does not know anything about the user’s LaTeX file! That’s the whole point: Aligning the document font size used by the PGF backend with `font.size` allows the user to control what font size is used internally by matplotlib, and **to set this to match their own document**! If we don’t do this, we have to arbitrarily hard-code a fixed document font size, making it **impossible** to include the generated PGF in LaTeX documents with other font sizes (there are workarounds of course, see #26892).

Now, I agree that the ideal solution for this problem would be for the user to be able to set the documentclass, at which point the code I’ve added which is setting the document font size could be removed. But I just wanted to fix the immediate problem to make things work *by default, without workarounds* in the custom preamble when making this PR!

**Comment by voidstarstar** [comment]
> Now, for @voidstarstar’s example: Of course, the assumption is that **the user will set `font.size` to match the font size of their document**! What scenario or use case would there be where any other choice makes sense?

This seems like the core of my confusion.  Please consider the following use case:

A user is writing a LaTeX paper where the plots are somewhat small for space reasons due to journal imposed page limitations.  The font size in the plots must be smaller than the document font size in order for the text to fit.  By default, all of the text element sizes (`axes.titlesize`, `axes.labelsize`, `xtick.labelsize`, `legend.fontsize`, etc.) use relative sizing (`large`, `medium`, etc.).  This means the user can change `font.size` and the entire figure will scale proportionally.  These relative sizes are converted to absolute pt sizes when writing the pgf file, so nothing is dependent upon the default document font size.

> > My recommended solution would be to leave the `\normalsize` and other commands (`\tiny`, etc.) alone to match the LaTeX file
> 
> How? matplotlib does not know anything about the user’s LaTeX file! That’s the whole point: Aligning the document font size used by the PGF backend with `font.size` allows the user to control what font size is used internally by matplotlib, and **to set this to match their own document**! If we don’t do this, we have to arbitrarily hard-code a fixed document font size, making it **impossible** to include the generated PGF in LaTeX documents with other font sizes (there are workarounds of course, see #26892).

The 2nd half of my sentence mentions using `pgf.preamble` which is what you did in your workaround.  The issue I have with making the workaround built in to matplotlib is that it breaks the use cases where `font.size` does not match the document size.

> Now, I agree that the ideal solution for this problem would be for the user to be able to set the documentclass, at which point the code I’ve added which is setting the document font size could be removed. But I just wanted to fix the immediate problem to make things work _by default, without workarounds_ in the custom preamble when making this PR!

I agree.  I'm personally not opposed to this as a temporary stepping stone solution.

**Comment by Socob** [comment]
> A user is writing a LaTeX paper where the plots are somewhat small for space reasons due to journal imposed page limitations. The font size in the plots must be smaller than the document font size in order for the text to fit. By default, all of the text element sizes (`axes.titlesize`, `axes.labelsize`, `xtick.labelsize`, `legend.fontsize`, etc.) use relative sizing (`large`, `medium`, etc.). This means the user can change `font.size` and the entire figure will scale proportionally. These relative sizes are converted to absolute pt sizes when writing the pgf file, so nothing is dependent upon the default document font size.

Setting the document font size doesn’t impair this use case at all. You can still set `font.size` to something other than the final LaTeX document’s font size; there is only a problem if you’re doing something that’s inherently tied to the document font size, such as using LaTeX font size commands in the plot or using a package like `unicode-math`. Clearly, by using these, you’re necessarily introducing a dependence on the document font size into your plot. Only in this case, you have to set `font.size` to match your document font size. And in this scenario, I do have a hard time imagining how any other setting could make sense.

Let’s also keep in mind that you’d *always* have this problem with a hard-coded document font size (unless your document’s font size just happened to match the matplotlib-internal one), so I don’t see any alternative!

**Comment by voidstarstar** [comment]
> Let’s also keep in mind that you’d _always_ have this problem with a hard-coded document font size (unless your document’s font size just happened to match the matplotlib-internal one), so I don’t see any alternative!

That's a very good point, I agree.  The hard-coded document font size is really the core issue, not your code.  Your code at least has the benefit of allowing the user to set `font.size` to have _some_ control over it.  Until the hard-coded `documentclass` is removed, there will always be _some_ use case that will be broken and the one you are targeting would also benefit more users.

Sorry about the confusion, you are correct.  I retract my example/problem mentioned above.  This PR has my :+1: (for whatever that's worth).

So it sounds to me like the best course of action would be:
1. First, #26893 (this PR) gets merged.
2. Then, the temporary `scrextend` code introduced by this PR should be removed in #28167 and replaced with a fully customizable `documentclass`.
3. Finally, #28167 gets merged.

**Comment by voidstarstar** [comment]
> Thank you @voidstarstar for your example. I agree we can't assume that the user will indeed be using koma-script just because it happens to be present in the install (if I understood the problem you're mentioning correctly).

@anntzer The problem I was mentioning was not really an issue with this PR, but rather an consequence of having a hard-coded `documentclass`.  As @Socob mentioned (and I agree with), the code in this PR does not assume koma-script is being used.  The command for conveniently changing the default sizes just happens to be part of that library, but seems to be often used on its own.  Please disregard my previous objection and example code.  My opinion is now LGTM!
